packages/react-storage/src/components/StorageBrowser/actions/handlers/utils.ts

```diff
@@ -0,0 +1,53 @@
+import { isCancelError } from 'aws-amplify/storage';
+import { isFunction } from '@aws-amplify/ui';
+
+import {
+  TaskHandlerOutput,
+  CancelableTaskHandlerOutput,
+  TaskHandlerOptions,
+  ActionInputConfig,
+} from '../types';
+
+interface TaskHandlerCallbacks
+  extends Pick<TaskHandlerOptions, 'onComplete' | 'onError'> {
+  onCancel?: (key: string) => void;
+}
+
+export const resolveHandlerResult = <T extends boolean>({
+  result,
+  key,
+  isCancelable,
+  options,
+}: {
+  result: Promise<any>;
+  key: string;
+  isCancelable: T;
+  options?: TaskHandlerCallbacks;
+}): (T extends false
+  ? TaskHandlerOutput
+  : CancelableTaskHandlerOutput)['result'] => {
+  const { onCancel, onComplete, onError } = options ?? {};
+  return result
+    .then(() => {
+      if (isFunction(onComplete)) onComplete(key);
+      return 'COMPLETE' as const;
+    })
+    .catch((error: Error) => {
+      if (isCancelable && isCancelError(error)) {
+        // const { onCancel } = options ?? {};
+        if (isFunction(onCancel)) onCancel(key);
```

Comment:
nit: (can ignore me) it can just be the following to save a few bytes towards the bundle?
```suggestion
        onCancel?.(key);
```

Reply:
Since `onCancel` might be provided by consumer prefer to ensure that `onCancel` is a function before calling